subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/OptionReader.java

```diff
@@ -185,7 +193,11 @@ private List<OptionElement> getFieldAnnotations(Class<?> type) {
         for (Field field : type.getDeclaredFields()) {
             Option option = findOption(field);
             if (option != null) {
-                fieldOptionElements.add(FieldOptionElement.create(option, field, optionValueNotationParserFactory));
+                OptionElement optionElement = FieldOptionElement.create(option, field, optionValueNotationParserFactory);
+                fieldOptionElements.add(optionElement);
+                if (optionElement.getOptionType().equals(Void.TYPE)) {
```

Comment:
:x: This test is looking at the option argument type. There are more options than `BooleanOptionElement` that have `void` argument type (e.g. `--rerun`). This might be the cause of the failures in `CommandLineTaskConfigurerSpec`. It would make more sense to test if `optionElement instanceOf BooleanOptionElement`.

But, that being said, I would say that the creation of the disabling options could be entirely moved out of `OptionReader` into `CommandLineTaskConfigurer`. Then the handling of clashing options and dis-allowance of mutually exclusive options could be simplified.